File: CartoMap.Objective-C/CartoMap.Objective-C/GalleryRow.m
//
//  GalleryRow.m
//  CartoMap.Objective-C
//
//  Created by Aare Undo on 28/09/2017.
//  Copyright © 2017 Aare Undo. All rights reserved.
//

#import "GalleryRow.h"

@implementation GalleryRow

- (id)init {
    
    self = [super init];
    
    self.backgroundColor = UIColor.whiteColor;
    
    self.layer.borderWidth = 1;
    self.layer.borderColor = [[UIColor colorWithRed:245/255.0f green:245/255.0f blue:245/255.0f alpha:1] CGColor];
    
    self.imageView = [[UIImageView alloc] init];
    self.imageView.contentMode = UIViewContentModeScaleAspectFill;
    self.imageView.clipsToBounds = YES;
    [self addSubview:self.imageView];
    
    self.titleView = [[UILabel alloc] init];
    self.titleView.textColor = [UIColor colorWithRed:14/255.0f green:122/255.0f blue:254/255.0f alpha:1];
    self.titleView.font = [UIFont fontWithName:@"HelveticaNeue-Bold" size:14];
    [self addSubview:self.titleView];
    
    self.descriptionView = [[UILabel alloc] init];
    self.descriptionView.textColor = [UIColor colorWithRed:100/255.0f green:100/255.0f blue:100/255.0f alpha:1];
    self.descriptionView.lineBreakMode = NSLineBreakByWordWrapping;
    self.descriptionView.numberOfLines = 0;
    self.descriptionView.font = [UIFont fontWithName:@"HelveticaNeue" size:12];
    [self addSubview:self.descriptionView];
    
    return self;
}

- (void)layoutSubviews {
    
    [self.titleView sizeToFit];
    [self.descriptionView sizeToFit];
    
    CGFloat padding = 5.0;
    CGFloat imageHeight = self.frame.size.height / 5 * 3;
    
    CGFloat x = padding;
    CGFloat y = padding;
    CGFloat w = self.frame.size.width - 2 * padding;
    CGFloat h = imageHeight;
    
    self.imageView.frame = CGRectMake(x, y, w, h);
    
    y += h + padding;
    h = self.titleView.frame.size.height;
    
    self.titleView.frame = CGRectMake(x, y, w, h);
    
    y += h + padding;
    h = self.descriptionView.frame.size.height;
    
    self.descriptionView.frame = CGRectMake(x, y, w, h);
    
    self.layer.masksToBounds = NO;
    self.layer.shadowColor = [UIColor colorWithRed:100/255 green:100/255 blue:100/255 alpha:1].CGColor;
    self.layer.shadowOffset = CGSizeMake(2.0, 2.0);
    self.layer.shadowOpacity = 0.5;
    self.layer.shadowPath = [UIBezierPath bezierPathWithRect:self.bounds].CGPath;
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    self.alpha = 0.5f;
    [super touchesBegan:touches withEvent:event];
}

- (void)touchesEnded:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    self.alpha = 1.0f;
    [super touchesEnded:touches withEvent:event];
}

- (void)touchesCancelled:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    self.alpha = 1.0f;
    [super touchesCancelled:touches withEvent:event];
}

- (void)update:(Sample *)sample {
    self.sample = sample;
    
    self.imageView.image = [UIImage imageNamed:sample.imageUrl];
    self.titleView.text = sample.title;
    self.descriptionView.text = sample.subtitle;
}

@end

File: AdvancedMap.Objective-C/AdvancedMap/GalleryRow.swift
//
//  GalleryRow.swift
//  Feature Demo
//
//  Created by Aare Undo on 19/06/2017.
//  Copyright © 2017 CARTO. All rights reserved.
//

import Foundation
import UIKit

@objc class GalleryRow : UIView {
    
    @objc var imageView: UIImageView!
    @objc var titleView: UILabel!
    @objc var descriptionView: UILabel!
    
    @objc var sample: Sample!
    
    convenience init() {
        
        self.init(frame: CGRect.zero)
        
        backgroundColor = UIColor.white
        
        layer.borderWidth = 1;
        layer.borderColor = Colors.nearWhite.cgColor
        
        imageView = UIImageView()
        imageView.contentMode = .scaleAspectFill
        imageView.clipsToBounds = true
        
        addSubview(imageView)
        
        titleView = UILabel()
        titleView.textColor = Colors.appleBlue
        titleView.font = UIFont(name: "HelveticaNeue-Bold", size: 14)
        
        addSubview(titleView)
        
        descriptionView = UILabel()
        descriptionView.textColor = UIColor.lightGray
        descriptionView.font = UIFont(name: "HelveticaNeue", size: 12)
        descriptionView.lineBreakMode = .byWordWrapping
        descriptionView.numberOfLines = 0
        
        addSubview(descriptionView)
    }
    
    override func layoutSubviews() {
        
        let padding: CGFloat = 5.0
        
        let imageHeight = frame.height / 5 * 3
        
        titleView.sizeToFit()
        descriptionView.sizeToFit()
        
        let titleHeight = titleView.frame.height
        let descriptionHeight = descriptionView.frame.height
        
        let x = padding
        var y = padding
        let w = frame.width - 2 * padding
        var h = imageHeight
        
        imageView.frame = CGRect(x: x, y: y, width: w, height: h)
        
        y += h + padding;
        h = titleHeight
        
        titleView.frame = CGRect(x: x, y: y, width: w, height: h)
        
        y += h + padding
        h = descriptionHeight
        
        descriptionView.frame = CGRect(x: x, y: y, width: w, height: h)
        
        layer.masksToBounds = false
        layer.shadowColor = UIColor.lightGray.cgColor
        layer.shadowOffset = CGSize(width: 2.0, height: 2.0)
        layer.shadowOpacity = 0.5
        layer.shadowPath = UIBezierPath(rect: bounds).cgPath
    }
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        alpha = 0.5
        super.touchesBegan(touches, with: event)
    }
    
    override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        alpha = 1.0
        super.touchesEnded(touches, with: event)
    }
    
    override func touchesCancelled(_ touches: Set<UITouch>, with event: UIEvent?) {
        alpha = 1.0
        super.touchesCancelled(touches, with: event)
    }
    
    @objc func update(sample: Sample) {
        
        self.sample = sample
        
        imageView.image = UIImage(named: sample.imageUrl)
        titleView.text = sample.title
        descriptionView.text = sample.subtitle
    }
}




